lcls_tools/common/data_analysis/projection_fit/projection_fit.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+import scipy.optimize
+from matplotlib import pyplot as plt
+from pydantic import BaseModel, ConfigDict
+from lcls_tools.common.data_analysis.projection_fit.method_base import MethodBase
+
+
+class ProjectionFit(BaseModel):
+    """
+    1d fitting class that allows users to choose the model with which the fit is performed,
+    and if prior assumptions (bayesian regression) about the data should be used when performing the fit.
+        Additionally there is an option to visualize the fitted data and priors.
+    -To perform a 1d fit, call fit_projection(projection_data={*data_to_fit*})
+    ------------------------
+    Arguments:
+    model: MethodBase (this argument is a child class object of method base
+        e.g GaussianModel & DoubleGaussianModel)
+    visualize_priors: bool (shows plots of the priors and init guess distribution before fit)
+    use_priors: bool (incorporates prior distribution information into fit)
+    visualize_fit: bool (visualize the parameters as a function of the forward function
+        from our model compared to distribution data)
+    """
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    model: MethodBase
+    visualize_priors: bool = False
+    use_priors: bool = False
+    visualize_fit: bool = False
+
+    def normalize(self, old_data: np.ndarray) -> np.ndarray:
+        """
+        Normalize a 1d array by scaling and shifting data
+        s.t. data is between 0 and 1
+        """
+        data = old_data.copy()
+        normalized_data = data / (np.max(data))
+        return normalized_data
+
+    def unnormalize_model_params(
+        self, params_dict: dict, projection_data: np.ndarray
```

Comment:
Ah yes, so for this we find fitted parameters to the normalized data. I could unnormalize the data as well but since we already have that I don't do that here. Instead I map normalized fit parameters back to what the fit would be on the normalized data.

Reply:
Ok, let's discuss this one in person too.